trie-db/src/lookup.rs

```diff
@@ -655,4 +668,219 @@ where
 		}
 		Ok(None)
 	}
+
+	/// Look up the merkle value (hash) of the node that is the closest descendant for the provided
+	/// key.
+	///
+	/// When the provided key leads to a node, then the merkle value of that node
+	/// is returned. However, if the key does not lead to a node, then the merkle value
+	/// of the closest descendant is returned. `None` if no such descendant exists.
+	fn inner_lookup_first_descendent(
+		mut self,
+		nibble_key: NibbleSlice,
+		full_key: &[u8],
+	) -> Result<Option<TrieHash<L>>, TrieHash<L>, CError<L>> {
```

Comment:
The return type is wrong.
```
enum MerkleValue {
     /// Is returned if `v.len() < MAX_INLINE_VALUE`.
     Node(Vec<u8>),
     Hash(H),
}
```

Reply:
I did propose to use hash for inline node (generally I would prefer if inline node is something users don't need to be aware off). But I did not double check the rpc proposal.

Reply:
Maybe trie crate api could return this enum, but the rpc return hash of the data value for inline node (this way we can avoid some hashing if this api get use by runtime in the future).

Reply:
But the spec is speaking about the merkle value and not the hash. The point being that if the data is smaller, we save bandwidth etc. 



> (this way we can avoid some hashing if this api get use by runtime in the future)

This doesn't add any extra hashing. If you need hashing for the runtime, we can just do it where we use it. There would be no extra hashing involved as we currently also hash the inline node manually.

Reply:
That makes sense! I've added the `MerkleValue` enum to be explicit about what the trie-db returns.

Then in substrate's RPC layer we could further `hash(MerkleValue::Node(..))` such that we return a consistent view of nodes and not care about inline ones. And if this function gets ever called from the runtime we save a `Layout:hash()` call 🙏

Reply:
> Then in substrate's RPC layer we could further `hash(MerkleValue::Node(..))` such that we return a consistent view of nodes and not care about inline ones.

This still goes against the size argument I laid out above. I opened https://github.com/paritytech/json-rpc-interface-spec/issues/88 to be sure that we use the correct value in the RPC.